src/categorical_algebra/DPO.jl

```diff
@@ -0,0 +1,215 @@
+module DPO
+export rewrite, rewrite_match, valid_dpo, dangling_condition, id_condition, pushout_complement, extend_cset,subcset
+
+using ..FinSets
+import ..CSets: AbstractACSet, ACSet, ACSetTransformation, components, homomorphism, homomorphisms, unpack_components, pack_components, add_parts!, nparts, set_subpart!, is_natural
+using ...Theories
+using ..Limits
+
+"""
+    l
+  L ← I
+m ↓   ↓k
+  G ← K
+    g
+
+Compute a pushout complement, componentwise in Set. On a formal level:
+For each component, define K = G / m(L/l(I)). There is a natural injection g: K↪G
+For each component, define k as equal to the map l;m (every element in the image in G is also in K).
+
+Returns ACSetTransformations k and g such that (m, g) is the pushout of (l, k)
+
+Implementation-wise, elements of K are ordered in the same order as they appear in G.
+Construct an offset mapping to keep track of the location of elements of G within K.
+"""
+function pushout_complement(l::ACSetTransformation,m::ACSetTransformation)::Pair{ACSetTransformation,ACSetTransformation}
+
+  I, L, G = l.dom, l.codom, m.codom
+  lm = compose(l, m)
+
+  k_components = Dict{Symbol, FinFunction}()
+  g_components = Dict{Symbol,Vector{Int}}()
+  K = ACSet{typeof(G).parameters...}()
+  offsets = Dict{Symbol,Vector{Int}}()
+
+
+  for comp in keys(l.components)
+
+    n_comp = nparts(G, comp)  # total # of elements in G
+    l_image = Set(l.components[comp].func)
```

Comment:
You never want to directly access the `func` field of a `FinFunctionVector` because other kinds of `FinFunctions` may have different fields.

So this should be `Set(collect(l))`.